// focus-lock/dist/es5/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var constants = tslib_1.__importStar(require("./constants"));
exports.constants = constants;
var focusables_1 = require("./focusables");
exports.getFocusabledIn = focusables_1.getFocusabledIn;
var focusInside_1 = require("./focusInside");
exports.focusInside = focusInside_1.focusInside;
var focusIsHidden_1 = require("./focusIsHidden");
exports.focusIsHidden = focusIsHidden_1.focusIsHidden;
var focusMerge_1 = require("./focusMerge");
exports.focusMerge = focusMerge_1.getFocusMerge;
var setFocus_1 = require("./setFocus");
var sibling_1 = require("./sibling");
exports.focusNextElement = sibling_1.focusNextElement;
exports.focusPrevElement = sibling_1.focusPrevElement;
var tabHook_1 = tslib_1.__importDefault(require("./tabHook"));
exports.tabHook = tabHook_1.default;
var all_affected_1 = require("./utils/all-affected");
exports.getAllAffectedNodes = all_affected_1.getAllAffectedNodes;
exports.default = setFocus_1.setFocus;
