// @authorizerdev/authorizer-js/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Authorizer: () => Authorizer,
  OAuthProviders: () => OAuthProviders,
  ResponseTypes: () => ResponseTypes
});
module.exports = __toCommonJS(src_exports);
var import_cross_fetch = __toESM(require("cross-fetch"));

// src/constants.ts
var DEFAULT_AUTHORIZE_TIMEOUT_IN_SECONDS = 60;
var CLEANUP_IFRAME_TIMEOUT_IN_SECONDS = 2;

// src/types.ts
var OAuthProviders;
(function(OAuthProviders2) {
  OAuthProviders2["Apple"] = "apple";
  OAuthProviders2["Github"] = "github";
  OAuthProviders2["Google"] = "google";
  OAuthProviders2["Facebook"] = "facebook";
  OAuthProviders2["LinkedIn"] = "linkedin";
  OAuthProviders2["Twitter"] = "twitter";
  OAuthProviders2["Microsoft"] = "microsoft";
  OAuthProviders2["Twitch"] = "twitch";
  OAuthProviders2["Roblox"] = "roblox";
  OAuthProviders2["Discord"] = "discord";
})(OAuthProviders || (OAuthProviders = {}));
var ResponseTypes;
(function(ResponseTypes2) {
  ResponseTypes2["Code"] = "code";
  ResponseTypes2["Token"] = "token";
})(ResponseTypes || (ResponseTypes = {}));

// src/utils.ts
var hasWindow = /* @__PURE__ */ __name(() => typeof window !== "undefined", "hasWindow");
var trimURL = /* @__PURE__ */ __name((url) => {
  let trimmedData = url.trim();
  const lastChar = trimmedData[trimmedData.length - 1];
  if (lastChar === "/")
    trimmedData = trimmedData.slice(0, -1);
  return trimmedData;
}, "trimURL");
var getCrypto = /* @__PURE__ */ __name(() => {
  return hasWindow() ? window.crypto || window.msCrypto : null;
}, "getCrypto");
var getCryptoSubtle = /* @__PURE__ */ __name(() => {
  const crypto = getCrypto();
  return crypto && crypto.subtle || crypto.webkitSubtle;
}, "getCryptoSubtle");
var createRandomString = /* @__PURE__ */ __name(() => {
  const charset = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-_~.";
  let random = "";
  const crypto = getCrypto();
  if (crypto) {
    const randomValues = Array.from(crypto.getRandomValues(new Uint8Array(43)));
    randomValues.forEach((v) => random += charset[v % charset.length]);
  }
  return random;
}, "createRandomString");
var encode = /* @__PURE__ */ __name((value) => hasWindow() ? btoa(value) : Buffer.from(value).toString("base64"), "encode");
var createQueryParams = /* @__PURE__ */ __name((params) => {
  return Object.keys(params).filter((k) => typeof params[k] !== "undefined").map((k) => `${encodeURIComponent(k)}=${encodeURIComponent(params[k])}`).join("&");
}, "createQueryParams");
var sha256 = /* @__PURE__ */ __name(async (s) => {
  const digestOp = getCryptoSubtle().digest({
    name: "SHA-256"
  }, new TextEncoder().encode(s));
  if (window.msCrypto) {
    return new Promise((resolve, reject) => {
      digestOp.oncomplete = (e) => {
        resolve(e.target.result);
      };
      digestOp.onerror = (e) => {
        reject(e.error);
      };
      digestOp.onabort = () => {
        reject(new Error("The digest operation was aborted"));
      };
    });
  }
  return await digestOp;
}, "sha256");
var urlEncodeB64 = /* @__PURE__ */ __name((input) => {
  const b64Chars = {
    "+": "-",
    "/": "_",
    "=": ""
  };
  return input.replace(/[+/=]/g, (m) => b64Chars[m]);
}, "urlEncodeB64");
var bufferToBase64UrlEncoded = /* @__PURE__ */ __name((input) => {
  const ie11SafeInput = new Uint8Array(input);
  return urlEncodeB64(window.btoa(String.fromCharCode(...Array.from(ie11SafeInput))));
}, "bufferToBase64UrlEncoded");
var executeIframe = /* @__PURE__ */ __name((authorizeUrl, eventOrigin, timeoutInSeconds = DEFAULT_AUTHORIZE_TIMEOUT_IN_SECONDS) => {
  return new Promise((resolve, reject) => {
    const iframe = window.document.createElement("iframe");
    iframe.setAttribute("id", "authorizer-iframe");
    iframe.setAttribute("width", "0");
    iframe.setAttribute("height", "0");
    iframe.style.display = "none";
    const removeIframe = /* @__PURE__ */ __name(() => {
      if (window.document.body.contains(iframe)) {
        window.document.body.removeChild(iframe);
        window.removeEventListener("message", iframeEventHandler, false);
      }
    }, "removeIframe");
    const timeoutSetTimeoutId = setTimeout(() => {
      removeIframe();
    }, timeoutInSeconds * 1e3);
    const iframeEventHandler = /* @__PURE__ */ __name(function(e) {
      if (e.origin !== eventOrigin)
        return;
      if (!e.data || !e.data.response)
        return;
      const eventSource = e.source;
      if (eventSource)
        eventSource.close();
      e.data.response.error ? reject(e.data.response) : resolve(e.data.response);
      clearTimeout(timeoutSetTimeoutId);
      window.removeEventListener("message", iframeEventHandler, false);
      setTimeout(removeIframe, CLEANUP_IFRAME_TIMEOUT_IN_SECONDS * 1e3);
    }, "iframeEventHandler");
    window.addEventListener("message", iframeEventHandler, false);
    window.document.body.appendChild(iframe);
    iframe.setAttribute("src", authorizeUrl);
  });
}, "executeIframe");

// src/index.ts
var userFragment = "id email email_verified given_name family_name middle_name nickname preferred_username picture signup_methods gender birthdate phone_number phone_number_verified roles created_at updated_at revoked_timestamp is_multi_factor_auth_enabled app_data";
var authTokenFragment = `message access_token expires_in refresh_token id_token should_show_email_otp_screen should_show_mobile_otp_screen should_show_totp_screen authenticator_scanner_image authenticator_secret authenticator_recovery_codes user { ${userFragment} }`;
var getFetcher = /* @__PURE__ */ __name(() => hasWindow() ? window.fetch : import_cross_fetch.default, "getFetcher");
var _Authorizer = class _Authorizer {
  // class variable
  config;
  codeVerifier;
  // constructor
  constructor(config) {
    if (!config)
      throw new Error("Configuration is required");
    this.config = config;
    if (!config.authorizerURL && !config.authorizerURL.trim())
      throw new Error("Invalid authorizerURL");
    if (config.authorizerURL)
      this.config.authorizerURL = trimURL(config.authorizerURL);
    if (!config.redirectURL && !config.redirectURL.trim())
      throw new Error("Invalid redirectURL");
    else
      this.config.redirectURL = trimURL(config.redirectURL);
    this.config.extraHeaders = {
      ...config.extraHeaders || {},
      "x-authorizer-url": this.config.authorizerURL,
      "x-authorizer-client-id": this.config.clientID || "",
      "Content-Type": "application/json"
    };
    this.config.clientID = ((config == null ? void 0 : config.clientID) || "").trim();
  }
  authorize = async (data) => {
    var _a;
    if (!hasWindow())
      return this.errorResponse([
        new Error("this feature is only supported in browser")
      ]);
    const scopes = [
      "openid",
      "profile",
      "email"
    ];
    if (data.use_refresh_token)
      scopes.push("offline_access");
    const requestData = {
      redirect_uri: this.config.redirectURL,
      response_mode: data.response_mode || "web_message",
      state: encode(createRandomString()),
      nonce: encode(createRandomString()),
      response_type: data.response_type,
      scope: scopes.join(" "),
      client_id: ((_a = this.config) == null ? void 0 : _a.clientID) || ""
    };
    if (data.response_type === ResponseTypes.Code) {
      this.codeVerifier = createRandomString();
      const sha = await sha256(this.codeVerifier);
      const codeChallenge = bufferToBase64UrlEncoded(sha);
      requestData.code_challenge = codeChallenge;
    }
    const authorizeURL = `${this.config.authorizerURL}/authorize?${createQueryParams(requestData)}`;
    if (requestData.response_mode !== "web_message") {
      window.location.replace(authorizeURL);
      return this.okResponse(void 0);
    }
    try {
      const iframeRes = await executeIframe(authorizeURL, this.config.authorizerURL, DEFAULT_AUTHORIZE_TIMEOUT_IN_SECONDS);
      if (data.response_type === ResponseTypes.Code) {
        const tokenResp = await this.getToken({
          code: iframeRes.code
        });
        return tokenResp.errors.length ? this.errorResponse(tokenResp.errors) : this.okResponse(tokenResp.data);
      }
      return this.okResponse(iframeRes);
    } catch (err) {
      if (err.error) {
        window.location.replace(`${this.config.authorizerURL}/app?state=${encode(JSON.stringify(this.config))}&redirect_uri=${this.config.redirectURL}`);
      }
      return this.errorResponse(err);
    }
  };
  browserLogin = async () => {
    try {
      const tokenResp = await this.getSession();
      return tokenResp.errors.length ? this.errorResponse(tokenResp.errors) : this.okResponse(tokenResp.data);
    } catch (err) {
      if (!hasWindow()) {
        return {
          data: void 0,
          errors: [
            new Error("browserLogin is only supported for browsers")
          ]
        };
      }
      window.location.replace(`${this.config.authorizerURL}/app?state=${encode(JSON.stringify(this.config))}&redirect_uri=${this.config.redirectURL}`);
      return this.errorResponse(err);
    }
  };
  forgotPassword = async (data) => {
    var _a;
    if (!data.state)
      data.state = encode(createRandomString());
    if (!data.redirect_uri)
      data.redirect_uri = this.config.redirectURL;
    try {
      const forgotPasswordResp = await this.graphqlQuery({
        query: "mutation forgotPassword($data: ForgotPasswordRequest!) {	forgot_password(params: $data) { message should_show_mobile_otp_screen } }",
        variables: {
          data
        }
      });
      return ((_a = forgotPasswordResp == null ? void 0 : forgotPasswordResp.errors) == null ? void 0 : _a.length) ? this.errorResponse(forgotPasswordResp.errors) : this.okResponse(forgotPasswordResp == null ? void 0 : forgotPasswordResp.data.forgot_password);
    } catch (error) {
      return this.errorResponse([
        error
      ]);
    }
  };
  getMetaData = async () => {
    var _a;
    try {
      const res = await this.graphqlQuery({
        query: "query { meta { version client_id is_google_login_enabled is_facebook_login_enabled is_github_login_enabled is_linkedin_login_enabled is_apple_login_enabled is_twitter_login_enabled is_microsoft_login_enabled is_twitch_login_enabled is_roblox_login_enabled is_email_verification_enabled is_basic_authentication_enabled is_magic_link_login_enabled is_sign_up_enabled is_strong_password_enabled is_multi_factor_auth_enabled is_mobile_basic_authentication_enabled is_phone_verification_enabled } }"
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse(res.data.meta);
    } catch (error) {
      return this.errorResponse([
        error
      ]);
    }
  };
  getProfile = async (headers) => {
    var _a;
    try {
      const profileRes = await this.graphqlQuery({
        query: `query {	profile { ${userFragment} } }`,
        headers
      });
      return ((_a = profileRes == null ? void 0 : profileRes.errors) == null ? void 0 : _a.length) ? this.errorResponse(profileRes.errors) : this.okResponse(profileRes.data.profile);
    } catch (error) {
      return this.errorResponse([
        error
      ]);
    }
  };
  // this is used to verify / get session using cookie by default. If using node.js pass authorization header
  getSession = async (headers, params) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: `query getSession($params: SessionQueryRequest){session(params: $params) { ${authTokenFragment} } }`,
        headers,
        variables: {
          params
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.session);
    } catch (err) {
      return this.errorResponse(err);
    }
  };
  getToken = async (data) => {
    if (!data.grant_type)
      data.grant_type = "authorization_code";
    if (data.grant_type === "refresh_token" && !data.refresh_token)
      return this.errorResponse([
        new Error("Invalid refresh_token")
      ]);
    if (data.grant_type === "authorization_code" && !this.codeVerifier)
      return this.errorResponse([
        new Error("Invalid code verifier")
      ]);
    const requestData = {
      client_id: this.config.clientID,
      code: data.code || "",
      code_verifier: this.codeVerifier || "",
      grant_type: data.grant_type || "",
      refresh_token: data.refresh_token || ""
    };
    try {
      const fetcher = getFetcher();
      const res = await fetcher(`${this.config.authorizerURL}/oauth/token`, {
        method: "POST",
        body: JSON.stringify(requestData),
        headers: {
          ...this.config.extraHeaders
        },
        credentials: "include"
      });
      const json = await res.json();
      if (res.status >= 400)
        return this.errorResponse([
          new Error(json.error_description || json.error)
        ]);
      return this.okResponse(json);
    } catch (err) {
      return this.errorResponse(err);
    }
  };
  login = async (data) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: `
					mutation login($data: LoginRequest!) { login(params: $data) { ${authTokenFragment}}}
				`,
        variables: {
          data
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.login);
    } catch (err) {
      return this.errorResponse([
        new Error(err)
      ]);
    }
  };
  logout = async (headers) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: " mutation { logout { message } } ",
        headers
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.response);
    } catch (err) {
      return this.errorResponse([
        err
      ]);
    }
  };
  magicLinkLogin = async (data) => {
    var _a, _b;
    try {
      if (!data.state)
        data.state = encode(createRandomString());
      if (!data.redirect_uri)
        data.redirect_uri = this.config.redirectURL;
      const res = await this.graphqlQuery({
        query: `
					mutation magicLinkLogin($data: MagicLinkLoginRequest!) { magic_link_login(params: $data) { message }}
				`,
        variables: {
          data
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.magic_link_login);
    } catch (err) {
      return this.errorResponse([
        err
      ]);
    }
  };
  oauthLogin = async (oauthProvider, roles, redirect_uri, state) => {
    let urlState = state;
    if (!urlState) {
      urlState = encode(createRandomString());
    }
    if (!Object.values(OAuthProviders).includes(oauthProvider)) {
      throw new Error(`only following oauth providers are supported: ${Object.values(oauthProvider).toString()}`);
    }
    if (!hasWindow())
      throw new Error("oauthLogin is only supported for browsers");
    if (roles && roles.length)
      urlState += `&roles=${roles.join(",")}`;
    window.location.replace(`${this.config.authorizerURL}/oauth_login/${oauthProvider}?redirect_uri=${redirect_uri || this.config.redirectURL}&state=${urlState}`);
  };
  resendOtp = async (data) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: `
					mutation resendOtp($data: ResendOTPRequest!) { resend_otp(params: $data) { message }}
				`,
        variables: {
          data
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.resend_otp);
    } catch (err) {
      return this.errorResponse([
        err
      ]);
    }
  };
  resetPassword = async (data) => {
    var _a, _b;
    try {
      const resetPasswordRes = await this.graphqlQuery({
        query: "mutation resetPassword($data: ResetPasswordRequest!) {	reset_password(params: $data) { message } }",
        variables: {
          data
        }
      });
      return ((_a = resetPasswordRes == null ? void 0 : resetPasswordRes.errors) == null ? void 0 : _a.length) ? this.errorResponse(resetPasswordRes.errors) : this.okResponse((_b = resetPasswordRes.data) == null ? void 0 : _b.reset_password);
    } catch (error) {
      return this.errorResponse([
        error
      ]);
    }
  };
  revokeToken = async (data) => {
    if (!data.refresh_token && !data.refresh_token.trim())
      return this.errorResponse([
        new Error("Invalid refresh_token")
      ]);
    const fetcher = getFetcher();
    const res = await fetcher(`${this.config.authorizerURL}/oauth/revoke`, {
      method: "POST",
      headers: {
        ...this.config.extraHeaders
      },
      body: JSON.stringify({
        refresh_token: data.refresh_token,
        client_id: this.config.clientID
      })
    });
    const responseData = await res.json();
    return this.okResponse(responseData);
  };
  signup = async (data) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: `
					mutation signup($data: SignUpRequest!) { signup(params: $data) { ${authTokenFragment}}}
				`,
        variables: {
          data
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.signup);
    } catch (err) {
      return this.errorResponse([
        err
      ]);
    }
  };
  updateProfile = async (data, headers) => {
    var _a, _b;
    try {
      const updateProfileRes = await this.graphqlQuery({
        query: "mutation updateProfile($data: UpdateProfileRequest!) {	update_profile(params: $data) { message } }",
        headers,
        variables: {
          data
        }
      });
      return ((_a = updateProfileRes == null ? void 0 : updateProfileRes.errors) == null ? void 0 : _a.length) ? this.errorResponse(updateProfileRes.errors) : this.okResponse((_b = updateProfileRes.data) == null ? void 0 : _b.update_profile);
    } catch (error) {
      return this.errorResponse([
        error
      ]);
    }
  };
  deactivateAccount = async (headers) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: "mutation deactivateAccount { deactivate_account { message } }",
        headers
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.deactivate_account);
    } catch (error) {
      return this.errorResponse([
        error
      ]);
    }
  };
  validateJWTToken = async (params) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: "query validateJWTToken($params: ValidateJWTTokenRequest!){validate_jwt_token(params: $params) { is_valid claims } }",
        variables: {
          params
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.validate_jwt_token);
    } catch (error) {
      return this.errorResponse([
        error
      ]);
    }
  };
  validateSession = async (params) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: `query validateSession($params: ValidateSessionRequest){validate_session(params: $params) { is_valid user { ${userFragment} } } }`,
        variables: {
          params
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.validate_session);
    } catch (error) {
      return this.errorResponse([
        error
      ]);
    }
  };
  verifyEmail = async (data) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: `
					mutation verifyEmail($data: VerifyEmailRequest!) { verify_email(params: $data) { ${authTokenFragment}}}
				`,
        variables: {
          data
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.verify_email);
    } catch (err) {
      return this.errorResponse([
        err
      ]);
    }
  };
  resendVerifyEmail = async (data) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: `
					mutation resendVerifyEmail($data: ResendVerifyEmailRequest!) { resend_verify_email(params: $data) { message }}
				`,
        variables: {
          data
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.resend_verify_email);
    } catch (err) {
      return this.errorResponse([
        err
      ]);
    }
  };
  verifyOtp = async (data) => {
    var _a, _b;
    try {
      const res = await this.graphqlQuery({
        query: `
					mutation verifyOtp($data: VerifyOTPRequest!) { verify_otp(params: $data) { ${authTokenFragment}}}
				`,
        variables: {
          data
        }
      });
      return ((_a = res == null ? void 0 : res.errors) == null ? void 0 : _a.length) ? this.errorResponse(res.errors) : this.okResponse((_b = res.data) == null ? void 0 : _b.verify_otp);
    } catch (err) {
      return this.errorResponse([
        err
      ]);
    }
  };
  // helper to execute graphql queries
  // takes in any query or mutation string as value
  graphqlQuery = async (data) => {
    var _a;
    const fetcher = getFetcher();
    const res = await fetcher(`${this.config.authorizerURL}/graphql`, {
      method: "POST",
      body: JSON.stringify({
        query: data.query,
        variables: data.variables || {}
      }),
      headers: {
        ...this.config.extraHeaders,
        ...data.headers || {}
      },
      credentials: "include"
    });
    const json = await res.json();
    if ((_a = json == null ? void 0 : json.errors) == null ? void 0 : _a.length) {
      return {
        data: void 0,
        errors: json.errors
      };
    }
    return {
      data: json.data,
      errors: []
    };
  };
  errorResponse = (errors) => {
    return {
      data: void 0,
      errors
    };
  };
  okResponse = (data) => {
    return {
      data,
      errors: []
    };
  };
};
__name(_Authorizer, "Authorizer");
var Authorizer = _Authorizer;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Authorizer,
  OAuthProviders,
  ResponseTypes
});
